okta/request_executor.py

```diff
@@ -48,52 +52,165 @@ def __init__(self, config, cache, http_client=None):
                 "SSWS "
                 f"{self._config['client']['token']}"
             )
+        else:
+            # OAuth
+            self._oauth = OAuth(self, self._config)
 
         self._http_client = HTTPClient({
             'requestTimeout': self._request_timeout,
             'headers': self._default_headers
         })
 
-    def create_request(self, method: str, url: str, body: dict = None,
-                       headers: dict = None):
+    async def create_request(self, method: str, url: str, body: dict = None,
+                             headers: dict = {}):
+        """
+        Creates request for request executor's HTTP client.
+
+        Args:
+            method (str): HTTP Method to be used
+            url (str): URL to send request to
+            body (dict, optional): Request body. Defaults to None.
+            headers (dict, optional): Request headers. Defaults to {}.
+
+        Returns:
+            dict, Exception: Tuple of Dictionary repr of HTTP request and
+            exception raised during execution
+        """
+        # Base HTTP Request
         request = {
             "method": method
         }
 
         # Build request
-        headers = self._default_headers
+        # Get predetermined headers and build URL
+        headers.update(self._default_headers)
         url = self._config["client"]["orgUrl"] + url
 
+        # OAuth
         if self._authorization_mode == "PrivateKey":
             # check if access token exists
             if self._cache.contains("OKTA_ACCESS_TOKEN"):
                 access_token = self._cache.get("OKTA_ACCESS_TOKEN")
-                headers.update({"Authorization": access_token})
             else:
-                # o/w create token
-                pass
+                # if not, make one
+                # Generate using private key provided
+                access_token, error = await self._oauth.get_access_token()
+                # return error if problem retrieving token
+                if error:
+                    return (None, error)
 
-        return request
+            # finally, add to header and cache
+            headers.update({"Authorization": f"Bearer {access_token}"})
```

Comment:
Question: Is the `f` a typo?

Reply:
No, _f_ strings in Python allow you to directly call variables with curly braces in the string`{}`